kingpin/actors/support/api.py

```diff
@@ -0,0 +1,338 @@
+        # the 'path' variables are passed all the way down the instantiation
+        # chain.
+        merged_kwargs = dict(self._kwargs.items() + kwargs.items())
+
+        return self.__class__(
+            name=name,
+            config=self._attrs[name],
+            client=self._client,
+            *args, **merged_kwargs)
+
+    method.__name__ = name
+    return method
+
+
+class RestConsumer(object):
+
+    _CONFIG = {}
+    _ENDPOINT = None
+
+    def __init__(self, name=None, config=None, client=None, *args, **kwargs):
+        """Initialize the RestConsumer object.
+
+        The generic RestConsumer object (with no parameters passed in) looks at
+        the self.__class__._CONFIG dictionary and dynamically generates access
+        methods for the various API methods.
+
+        The GET, PUT, POST and DELETE methods optionally listed in
+        CONFIG['http_methods'] represent the possible types of HTTP methods
+        that the CONFIG['path'] supports. For each one of these listed, a
+        @coroutine wrapped get/put/post/delete() method will be created in the
+        RestConsumer that knows how to make the HTTP request.
+
+        For each item listed in CONFIG['attrs'], an access method is created
+        that will dynamically create and return a new RestConsumer object thats
+        configured for this endpoint. These methods are not asynchronous, but
+        are non-blocking.
+
+        Args:
+            name: Name of the resource method (default: None)
+            config: The dictionary object with the configuration for this API
+                    endpoint call.
+            client: <TBD>
+            *args,**kwargs: <TBD>
+        """
+        # If these aren't passed in, then get them from the class definition
+        name = name or self.__class__.__name__
+        config = config or self._CONFIG
+
+        # Get the basic options for this particular REST endpoint access object
+        self._path = config.get('path', None)
+        self._http_methods = config.get('http_methods', None)
+        self._attrs = config.get('attrs', None)
+        self._kwargs = kwargs
+
+        # If no client was supplied, then we
+        self._client = client or RestClient()
+
+        # Ensure that any tokens that need filling-in in the self._path setting
+        # are pulled from the **kwargs passed into this init. This is used on
+        # API paths like Hipchats '/v2/room/%(res)/...' URLs.
+        self._path = self._replace_path_tokens(self._path, kwargs)
+
+        # Create all of the methods based on the self._http_methods and
+        # self._attrs dict.
+        self._create_methods()
+        self._create_attrs()
+
+    def __repr__(self):
+        return '%s(%s)' % (self.__class__.__name__, self)
+
+    def __str__(self):
+        return str(self._path)
+
+    def __cmp__(self, other):
+        return cmp(self._path, other._path)
+
+    def _replace_path_tokens(self, path, kwargs):
```

Comment:
Using `kwargs` when not receiving `**kwargs` is confusing. Overall this function doesn't do anything except convert LookupError into a TypeError. It's not leveraging "self" so it's not class specific. I recommend deleting it altogether.